### ISSUE TITLE
Clarify rationale

### ISSUE THREAD

**Issue by ScottFreeCode** [MaxMotovilov/cort#1]
This is a really clever project and the opening problem statement would benefit from a more specific use case.

The current rationale section in the readme sounds as though merely being asynchronous causes code to explode in possible combinations of effects, but typically if one writes steps in terms of input and output (not side effects or shared state) and uses a systematic method of ensuring that steps with output needed for other steps' input run before those other steps (eg. promises) then no such interference between asynchronous steps occurs, let alone combinations thereof -- instead, steps are unordered with respect to others with which they do not interact and correctly ordered with respect to others with which they interact through input and output.

I'm wondering: Is this about testing changes to an inherently global resource such as the DOM or a database backend? A roundabout way of checking for side effects in otherwise well-defined asynchronous code, by proxy of whether those side effects happen to interfere with each other? A search for hidden race conditions? A means of wrangling imperative side-effect-based code that crashed headlong into the obstacles of (soft) concurrency (in the form of asynchronicity)? Or something else that just happens to be slipping my mind at the moment?

Relatedly, could the code examples be somewhat more comprehensive and/or realistic? Right now they basically just convey the bare minimum understanding of what the API is and the general fact that said API can be used to try multiple steps in multiple combinations of orders. Based on our conversation elsewhere I'm getting the sense that I'm missing something more to how it works in practice...

Thanks in advance!

**Comment by MaxMotovilov** [comment]
> if one writes steps in terms of input and output (not side effects or shared state) and uses a systematic method of ensuring that steps with output needed for other steps' input run before those other steps (eg. promises) then no such interference between asynchronous steps occurs

Indeed! This extension of the model is what I am working on right now, roughly speaking I'd like to support:

    later( ready => foo().then( x => (ready(), x) ) )

Since this code looks clunky I will certainly have some syntactic sugar; the hard part is correctly incorporating those "extended actions" (as opposed to just events) into the search tree without violating constraints imposed by the code (i.e. A goes after B by design, not happenstance) or missing viable permutations, such as this code misses now:

    let n=2;
    later( () => setTimeout( next( "A" ), 50 ) );
    later( () => setTimeout( next( "B" ), 100 ) );
    function next( tag ) { later( tag, () => --n || done() ) }

There is yet another elephant in the room that I have not yet decided how to deal with: non-deterministic behavior, mostly related to unpredictable timing. If it does not result in varied code paths (example above but with random delays) it is not a problem, if it does then potentially one may not be able to walk the same path twice! Current code detects this, but can't do much better than barf.

> Based on our conversation elsewhere I'm getting the sense that I'm missing something more to how it works in practice...

No doubt about it -- this is barely work in progress. I have started this project couple weeks ago when I wanted an exhaustive set of tests for an asynchronous message-processing engine and realized that none of the popular unit test frameworks provide this sort of functionality. Hence my attention towards point-in-time events: that's all I needed there so far. Now this whole idea has taken a life of its own so I'd certainly like to bring it to the point where it's both widely usable and comprehensible.

As to what sort of issues this approach is intended to detect -- I would say mostly unintended interference between side effects, of which race conditions I believe to be an example. I tend to think about it in slightly different terms: trying to approach a formal proof of correctness by demonstrably covering the entire state space [of the code in question] with tests. I would be the first to admit this statement is less than rigorous, but that's where I'm coming from, if it helps. Part of that is normally done by varying input data; I'd like to also vary relative timings which are known from experience to have major impact.
